### ISSUE TITLE
Google cloud functions?

### ISSUE THREAD

**Issue by zackify** [apollographql/apollo-server#357]
Can I use the lambda package and run it on google cloud? I'm looking into using google cloud functions with the project. Sorry I didn't investigate first. Just wanted to see if anyone else is interested as well!

**Comment by soda0289** [comment]
The AWS Lambda integration is designed to export a function for API Gateway proxy integration not Google Cloud. You would be better off using the graphql-server-core package and converting the request object into one that can be read by that package.

Is google cloud functions still in beta? It might be too early to write an integration for it.

**Comment by zackify** [comment]
Thanks for the info. Yes it is still in beta

**Comment by helfer** [comment]
I'd also be happy to accept a PR that adds a separate package for google cloud functions, just like we have one for lambda. :wink:

**Comment by zackify** [comment]
Is there anywhere I should look to better understand the internals? Or just look over the current code to make my own :)

**Comment by helfer** [comment]
Yeah, I would just look at what the lambda or micro package does and then roll your own! It's pretty straightforward.

**Comment by DxCx** [comment]
@zackify the http logic itself is pretty abstract in `graphql-core`
you can take a quick look at the integrations themselves and see that they are just "adapters".
should be pretty easy to add support..

**Comment by jthegedus** [comment]
I have also been looking into hosting GraphQL on GCP Cloud Functions / Cloud Functions for Firebase. I've come across a number of resources that have confused me with the use of ExpressJS in the context of Cloud Functions:

[Here](https://github.com/firebase/firebase-functions/issues/27) I see people passing an ExpressJS server into the Firebase Cloud Functions ``functions.https.onRequest`` with a [Firebase-Samples](https://github.com/firebase/functions-samples) example using the same method [here](https://github.com/firebase/functions-samples/blob/master/authorized-https-endpoint/functions/index.js).

But then I came across a single person [here](https://github.com/firebase/functions-samples/issues/101) who has concerns about creating an express app in a Cloud Function.

So I'm not sure which method here is correct in the context of ExpressJS in Cloud Functions for Firebase. I've [posted on SO](http://stackoverflow.com/questions/43579442/cloud-functions-for-firebase-and-express) with the following questions:
*   How does Cloud Functions for Firebase or GCP Cloud Functions handle the lifetime of objects initialised outside of the function definition?
*   How does the above effect the lifetime of the function? Does it run until timeout or function similarly to AWS Lambda?
*   The ``functions.https.onRequest`` documentation doesn't specify you can pass an ExpressJS server object into it, so how is this working? Are their then two endpoints? Can someone explain what is happening here?

**TLDR; if you can use ExpressJS server as described [here](https://github.com/firebase/functions-samples/blob/master/authorized-https-endpoint/functions/index.js) in Google/Firebase Cloud Functions could we then not just rely on graphql-server-express?**

**Comment by soda0289** [comment]
It does seem like the google cloud function takes in an express Request/Response object.
https://cloud.google.com/functions/docs/writing/http

It also seems like the firebase google cloud function has several triggers. The http trigger is created using this function `functions.https.onRequest`. Which takes in the same express Request and Response objects.
https://firebase.google.com/docs/functions/http-events

Does seem like the implementation of `graphql-server-express` might work without modification.

Not sure how to answer the other questions about function lifetime still learning about GCP.

**Comment by jthegedus** [comment]
I can confirm that the ``graphql-server-express`` is easily setup in Cloud Functions. It's more stable than the ``express-graphql`` implementation as it doesn't appear to error on the absent trailing '/'.

Sample code:
```js

// graphql-server-express
const bodyParser = require("body-parser");
const graphqlExpress = require("graphql-server-express").graphqlExpress;
const mySchema = ...;
var graphQLServer = express();
graphQLServer.use("*", bodyParser.json(), graphqlExpress({ schema: mySchema }));

// CF for Firebase with graphql-server-express
exports.graphql = functions.https.onRequest(graphQLServer);
```
called using ``https://us-central1-<project-name>.cloudfunctions.net/graphql``.

however to use GraphiQL as well you must do the following:
```js
// import graphiql
const graphiqlExpress = require("graphql-server-express").graphiqlExpress;

// assign specific URLs
graphQLServer.use("/graphql", bodyParser.json(), graphqlExpress({ schema: mySchema }));
graphQLServer.use("/graphiql", graphiqlExpress({ endpointURL: "/graphql" }));
```
and then call the method with ``https://us-central1-<project-name>.cloudfunctions.net/graphql/graphql`` and ``https://us-central1-<project-name>.cloudfunctions.net/graphql/graphiql``.

I still do not have an answer for how definitions of objects outside the function definition are handled or how Cloud Functions can take an entire ExpressJS object and just **magically work** seemingly stripping the object back to it's ``request, response`` parts. I've asked a detailed question on [SO](http://stackoverflow.com/questions/43579442/cloud-functions-for-firebase-and-express)

**Comment by zackify** [comment]
Also, how would you do this with subscriptions? Is that not possible I'm guessing?

**Comment by jthegedus** [comment]
[This](https://github.com/apollographql/graphql-subscriptions/issues/53) ``graphql-subscription`` issue is tracking that. I'm looking into managed PubSub services, particularly [Google Cloud PubSub](https://cloud.google.com/pubsub/) at the moment.

**Comment by stubailo** [comment]
@jthegedus just to close the loop, would you mind sending a PR to the README and docs [here](http://dev.apollodata.com/tools/apollo-server/setup.html) adding the code snippet about google cloud functions?

**Comment by jthegedus** [comment]
@stubailo It would be my pleasure :smile:

**Comment by Paddy-Hamilton** [comment]
@stubailo  @jthegedus  did an official best practice example ever make the docs? I presume not as I can't seem to find it in in the [server docs](https://www.apollographql.com/docs/apollo-server/)

Are their any plans to add one? 

I have used the the the example walk through in [this article](https://codeburst.io/apollo-graphql-server-express-google-cloud-function-serverless-framework-dbb6b5cf4bb3) and it worked well. Could I suggest reviewing it and maybe adding it to the docs ? 

let me know if there is any way I could help with that.  😄

**Comment by jthegedus** [comment]
@Paddy-Hamilton I did record this early Jan in my backlog of things to do. I should have time this weekend to address this. I'll ping you in the PR if you wish to give feedback. (that example you linked does actually reference [my article](https://codeburst.io/graphql-server-on-cloud-functions-for-firebase-ae97441399c0) for doing this on Firebase, I just never got around to vanilla Google Cloud Functions)

**Comment by Paddy-Hamilton** [comment]
@jthegedus ha, so it does, thanks for the help 👍 

Sounds good, will take a look at the PR, although i'm a relative graphql and GCF newbie so my feedback maybe limited. I'll help in whatever way I can.
Cheers 🍻
